server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/SpanServiceElasticsearch.java

```diff
@@ -192,6 +201,83 @@ public void clear(String tenantId) {
         }
     }
 
+    @Override
+    public Trace getTraceFragment(String tenantId, String id) {
+        Span span = getSpan(tenantId, id);
+
+        Trace trace = spanToTrace(span);
+        LinkedList<Node> nodes = recursiveTraceFragment(tenantId, id);
+        nodes.push(spanToNode(span));
+        trace.setNodes(nodes);
+
+        return trace;
+    }
+
+
+    private LinkedList<Node> recursiveTraceFragment(String tenantId, String spanId) {
+        List<Span> spanChildren = getChildren(tenantId, spanId);
+        if (spanChildren == null) {
+            return new LinkedList<>();
+        }
+
+        LinkedList<Node> nodes = new LinkedList<>();
+
+        for (Span span: spanChildren) {
+            if (span.serverSpan()) {
+                continue;
+            }
+
+            InteractionNode node = spanToNode(span);
+            node.setNodes(recursiveTraceFragment(tenantId, span.getId()));
+            nodes.add(node);
+        }
+
+        return nodes;
+    }
+
+    private Trace spanToTrace(Span span) {
+        if (span == null) {
+            throw new NullPointerException();
+        }
+
+        List<Property> spanProperties = span.properties();
+
+        Trace trace = new Trace();
+        trace.setId(span.getId());
+        trace.setStartTime(span.getTimestamp());
+        trace.setHostAddress(Span.ipv4Address(spanProperties));
+        trace.setProperties(new HashSet<>(span.properties()));
+
+//        TODO remove
+//        trace.setBusinessTransaction()
+//        trace.setHostName()
+//        trace.setPrincipal()
+
+        return trace;
+    }
+
+    private InteractionNode spanToNode(Span span) {
+        String endpointType = "http";
+        String url = span.url() != null ? span.url().toString() : null;
+
+        InteractionNode node;
+        if (span.serverSpan()) {
+            node = new Consumer(url, endpointType);
+        }else if (span.clientSpan()) {
+            node = new Producer(url, endpointType);
+        } else {
+            node = new Component(url, endpointType);
```

Comment:
Second parameter for Component is component type - which at the moment we don't know, so could just leave as null.